baseplate/frameworks/queue_consumer/kombu.py

```diff
@@ -124,22 +162,110 @@ def __init__(
+            logger.exception(
+                "Recoverable error while trying to process a message. "
+                "The message has been returned to the queue broker."
+            )
+            message.requeue()
+            return
+
+        headers = message.headers or {}
+        retry_count_val = headers.get(MESSAGE_HEADER_RETRY_COUNT, 0)
+        try:
+            retry_count = int(retry_count_val)
+        except (ValueError, TypeError):
+            retry_count = 0
+
+        retry_limit_val = headers.get(MESSAGE_HEADER_RETRY_LIMIT, None)
+        retry_limit: Optional[int]
+        try:
+            retry_limit = int(retry_limit_val)
+        except (ValueError, TypeError):
+            retry_limit = None
+
+        if (self.retry_limit is not None and retry_count >= self.retry_limit) or (
+            retry_limit is not None and retry_count >= retry_limit
+        ):
+            logger.exception(
+                "Unhandled error while trying to process a message. "
+                "The message reached the retry limit."
+            )
+            AMQP_REJECTED_TOTAL.labels(
+                **prometheus_labels._asdict(), reason_code=AMQP_REJECTED_REASON_RETRIES
+            ).inc()
+            message.reject()
+            return
+
+        headers[MESSAGE_HEADER_RETRY_COUNT] = retry_count + 1
+        message.ack()
+
+        new_message = message.channel.prepare_message(
+            message.body,
+            content_type=message.content_type,
+            content_encoding=message.content_encoding,
+            headers=headers,
+        )
+
+        message.channel.basic_publish(new_message, message_exchange, message_routing_key)
+        AMQP_REPUBLISHED_TOTAL.labels(**prometheus_labels._asdict()).inc()
+        logger.exception(
+            "Unhandled error while trying to process a message. "
+            "The retry message has been published to the queue broker."
+        )
+
+    def _is_ttl_over(self, message: kombu.Message) -> bool:
+        ttl = (message.headers or {}).get(MESSAGE_HEADER_TTL, 0)
+        return ttl and ttl < time.time()
 
     def handle(self, message: kombu.Message) -> None:
         start_time = time.perf_counter()
         prometheus_success = "true"
         prometheus_labels = AmqpConsumerPrometheusLabels(
-            amqp_address=message.channel.connection.client.host,  # note: localhost will be translated to 127.0.0.1 by the library
+            # note: localhost will be translated to 127.0.0.1 by the library
+            amqp_address=message.channel.connection.client.host,
             amqp_virtual_host=message.channel.connection.client.virtual_host,
             amqp_exchange_name=message.delivery_info.get("exchange", ""),
             amqp_routing_key=message.delivery_info.get("routing_key", ""),
         )
 
+        if self._is_ttl_over(message):
+            message.reject()
+            AMQP_REJECTED_TOTAL.labels(
```

Comment:
This applies to all messages not just AMQP ones. Is this intentional?

Reply:
Yes, it's intentional. Literally 12 lines below:  
```
AMQP_ACTIVE_MESSAGES.labels(
                **prometheus_labels._asdict()
            ).track_inprogress()
```  
Or, a bit more below:
```
            AMQP_PROCESSED_TOTAL.labels(
                **prometheus_labels._asdict(), amqp_success=prometheus_success
            ).inc()
```
I.e. as far as I can see this change extends an existing functionality, and not creating new conventions.